cdap-explore/src/main/java/co/cask/cdap/explore/executor/QueryResultsBodyProducer.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.explore.executor;
+
+import co.cask.cdap.explore.service.ExploreException;
+import co.cask.cdap.explore.service.ExploreService;
+import co.cask.cdap.explore.service.HandleNotFoundException;
+import co.cask.cdap.proto.QueryHandle;
+import co.cask.cdap.proto.QueryResult;
+import co.cask.http.BodyProducer;
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.SQLException;
+import java.util.List;
+
+/**
+ * BodyProducer used for returning the results of a Query, chunk by chunk.
+ */
+final class QueryResultsBodyProducer extends BodyProducer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(QueryResultsBodyProducer.class);
+
+  private final ExploreService exploreService;
+  private final QueryHandle handle;
+
+  private final StringBuffer sb;
```

Comment:
Instead of constructing string and copying into `byte[]`, better use `OutputStreamWriter` that wraps around `ChannelBufferOutputStream` then wraps around `ChannelBuffer`. Something like this.

```java
private final ChannelBuffer buffer;
private final PrintWriter writer;
....
// constructor
this.buffer = ChannelBuffers.dynamicBuffer();
this.writer = new PrintWriter(new OutputStreamWriter(new ChannelBufferOutputStream(buffer), StandardCharsets.UTF_8));
...
// In nextChunk()
buffer.clear();
if (result == null) {
  writer.println(getCSVHeaders(....));
  result = ...
}
for (QueryResult result : results) {
  ...
  writer.println(...);
}
writer.flush();
return buffer;
```